File: go/v2/run/pipe/pull/pull.go
package main

import (
	"fmt"
	"log"
	"os"
	"strconv"
	"sync"
	"time"

	protopipes "github.com/nochte/pipelinr-protocol/protobuf/pipes"

	pipes "github.com/nochte/pipelinr-clients/go/v2/pipe/http"
)

func main() {
	step := os.Getenv("STEP")
	threadstr := os.Getenv("THREADS")
	threads := 1
	if threadstr != "" {
		threads, _ = strconv.Atoi(threadstr)
	}
	url := os.Getenv("PIPELINR_URL")
	apikey := os.Getenv("PIPELINR_API_KEY")

	st := time.Now()
	for j := 0; j < threads; j++ {
		go func(workernum int) {
			receiver := pipes.New(url, "2", step, apikey)
			receiver.SetReceiveOptions(&protopipes.ReceiveOptions{
				AutoAck:           false,
				Block:             true,
				Count:             10,
				Timeout:           10,
				Pipe:              step,
				RedeliveryTimeout: 300,
			})
			go func() {
				if er := receiver.Start(); er != nil {
					panic(fmt.Sprintf("failed to start receiver %v\n", er.Error()))
				}
			}()

			time.Sleep(time.Second)

			ch := receiver.Chan()

			i := 0
			for m := range ch {
				if er := receiver.Log(m.GetStringId(), int32(i), fmt.Sprintf("finished iteration %v", i)); er != nil {
					log.Printf("error logging %v\n", er.Error())
				}
				if er := receiver.Decorate(m.GetStringId(), []*protopipes.Decoration{
					{Key: fmt.Sprintf("worker-%v", workernum), Value: fmt.Sprintf("%v", i)},
				}); er != nil {
					log.Printf("error logging %v\n", er.Error())
				}
				// if er := receiver.Ack(m.GetStringId()); er != nil {
				// 	log.Printf("error acking %v\n", er.Error())
				// }
				time.Sleep(time.Millisecond * 100)
				if er := receiver.Complete(m.GetStringId()); er != nil {
					log.Printf("error completing %v\n", er.Error())
				}

				i++
				if i%1000 == 0 {
					log.Printf("(%v) (%v) %v messages per sec \n", time.Since(st), i, float64(i)/float64(time.Since(st).Seconds()))
				}

			}
		}(j)
	}
	wg := sync.WaitGroup{}
	wg.Add(1)
	wg.Wait()
}

File: js/run/pipe/pull.js
const sleep = require('../../lib/sleep');
const HTTPPipe = require('../../pipe/http');

const dopipe = async (index) => {
	const step = process.env.STEP
  const url = process.env.PIPELINR_URL;
  const apikey = process.env.PIPELINR_API_KEY;

  const receiver = new HTTPPipe(apikey, step, url);
  receiver.setReceiveOptions({count: 10, timeout: 10, redeliveryTimeout: 300});
  receiver.start();

  for (let i = 1; i > 0; i++) {
    const m = await receiver.next();
    await receiver.log(m.id.value, i, `finished iteration ${i}`);
    await receiver.decorate(m.id.value, [{
      Key: `worker-${index}`,
      Value: `"${i}"`,
    }]);
    await receiver.complete(m.id.value);
    console.log(index, 'done with msg', m.id.value);
  }
}

const go = async () => {
  let threads = 1;
  if (process.env.THREADS) {
    threads = parseInt(process.env.THREADS);
  }
  if (threads === NaN || threads < 1) {
    threads = 1;
  }
  console.log('launching threads', threads);
  for (let i = 0; i < threads; i++) {
    dopipe(i);
  }

  while (true) {
    await sleep(10000);
  }
};

go();